### ISSUE TITLE
Always show colorcolumn if a line crosses it

### ISSUE THREAD

**Issue by Hubro** [Bekaboo/deadcolumn.nvim#12]
This plugin works really freaking well, I'm always a fan of reducing visual noise, but one this I miss is that I can no longer tell of a line in the current view crosses the column.

If I set `scope` to `"visible"`, then the point of the plugin is largely lost, I don't want to see the colorcolumn as long as nothing crosses it.

What would be perfect is if the plugin did exactly what it currently does, but if a currently visible line crosses the colorcolumn, it should be displayed no matter what.

Unless I've misunderstood the documentation, this isn't currently possible. Is there a elegant way this can be implemented?

---

Maybe something like:

```
warningscope = "buffer" | "visible"
```

Works like scope, but overrides it if a line actually crosses the colorcolumn. My ideal behavior would be `scope = "line"` and `warningscope = "visible"`. I would also probably like `warningscope = "buffer"` so I would immediately know if *any* line in the current file was too long.

**Comment by Bekaboo** [comment]
Hi, thanks for reporting.

> but one this I miss is that I can no longer tell of a line in the current view crosses the column.
> if a currently visible line crosses the colorcolumn, it should be displayed no matter what.

Seems like that setting `scope` to `visible` does exactly what you want.

**Comment by Hubro** [comment]
@Bekaboo Except for:

> If I set scope to "visible", then the point of the plugin is largely lost, I don't want to see the colorcolumn as long as nothing crosses it.

If I set scope to "visible", I pretty much *always* see the colorcolumn, because I almost *always* have at least one line in the current viewport that comes close to the colorcolumn.

**Comment by Bekaboo** [comment]
OK, I understand. Can you try setting `opts.blending.threshold` to `1` to show the colorcolumn in red only when a line is longer than `cc`

**Comment by Hubro** [comment]
@Bekaboo But that wouldn't be the default behavior, I like the default behavior :smile:

What I'm hoping for is to *expand* the default behavior to additionally show the colorcolumn when a line in the current viewport crosses it. If no lines in the current viewport crosses the colorcolumn, then the default behavior is perfect.

Setting `opts.blending.threshold` to `1` would (if I understand correctly) completely change the default behavior so that I only see the colorcolumn when I cross it, even in insert mode.

**Comment by Bekaboo** [comment]
Setting `opts.blending.threshold` to `1` should show the color column in red only when there exists a line in scope longer than `cc`.

> Setting opts.blending.threshold to 1 would (if I understand correctly) completely change the default behavior so that I only see the colorcolumn when I cross it, even in insert mode.

Whether the colored column will be shown in insert mode is controlled by `opts.modes`, if you do not want it to show in insert mode, you can set `opts.modes` to `{ 'n' }`.

If this is not what you expected, please elaborate, and express your desire in clearer way, e.g. using a table or a list to show in which modes, under what conditions, with what scope do you like to see the colored column. I have to admit that I don't fully get your point and your feature request seems vague to me.

**Comment by Hubro** [comment]
I think the default behavior is perfect as long as no lines are currently crossing the colorcolumn:

![image](https://user-images.githubusercontent.com/597206/235323724-2f07fb20-9a68-48c2-a20a-a1a8df3955bf.png)

![image](https://user-images.githubusercontent.com/597206/235323730-e2f3bf22-c3fc-41ae-8775-65b64704dba8.png)

However, as soon as a line actually crosses the colorcolumn, this is no longer so nice:

![image](https://user-images.githubusercontent.com/597206/235323760-5165925f-d909-49e8-86d0-9e965d4d668e.png)

Because now it's not clear if any lines are too long, and I have to actually move the cursor to each line to check if they are too long:

![image](https://user-images.githubusercontent.com/597206/235323822-f2ad9331-8f1b-4189-ab39-e8a8bafa759b.png)

Basically, I wish it was possible to keep `scope` at `"line"`, but have it behave as if it was set to `"visible"` only when a line actually crosses a colorcolumn in the current viewport.

My tentative suggestion was:

```lua
-- Default behavior
scope = "line"
```

```lua
-- Default behavior, but make colorcolumn visible if any lines in the current viewport
-- cross the colorcolumn
scope = "line"
warningscope = "visible"
```

```lua
-- Default behavior, but make colorcolumn visible if any lines in the current buffer
-- cross the colorcolumn
scope = "line"
warningscope = "buffer"
```

**Comment by Bekaboo** [comment]
Firstly, the images shown in the screenshots do not reflect the default behavior -- by default this plugin will not show the colored column in normal mode not matter what, so I assume you passed some custom config and let deadcolumn.nvim show to colored column in normal mode.

Secondly, if I do not misunderstand your proposal, you want the colored column to be shown in red if there is a line visible that is longer than `cc`, but if there is not such line, you would like the deadcolumn.nvim to behave as if it is configured with `opts.modes = { 'n' }`, right?

If that is what you expected, you can change the config dynamically when you scroll the screen, something like this might help:
. If you have any questions please fell free to ask me.
```lua
vim.api.nvim_create_autocmd({ 'VimResized', 'WinScrolled', 'BufWinEnter' }, {
    group = vim.api.nvim_create_augroup('DeadcolumnConfig', {}),
    callback = function()
        local lines = vim.api.nvim_buf_get_lines(
          0,
          vim.fn.line('w0') - 1,
          vim.fn.line('w$'),
          false
        )
        if math.max(unpack(vim.tbl_map(vim.fn.strdisplaywidth, lines))) > vim.wo.cc then
            require('deadcolumn').setup({
                scope = 'visible',
            })
        else
            require('deadcolumn').setup({
                scope = 'line',
            })
        end
    end
})
```

I have not tested it so there might be bugs, but you should get the point and modify the snippet as you wish. If you have any questions please fell free to ask me.

**Comment by Hubro** [comment]
You're right, I forgot I added "n" to the default modes, but fortunately that has nothing to do with my proposal.

My proposal is only to change the effective value of `scope` based on whether any currently visible lines cross the colorcolumn, which seems to be exactly what your snippet demonstrates :slightly_smiling_face:

I think it would be nice if this was possible to achieve with the options of the plugin. My first two thoughts after installing this plugin was:

1. Aah, fancy! My editor looks much cleaner now!
2. Oh dear, I wonder if any of my lines are too long, I have no idea...

Having a `scope` and a `warningscope` would let me have the best of both worlds. If I don't see the colorcolumn, I can know for sure none of my lines are too long :slightly_smiling_face:

**Comment by Bekaboo** [comment]
I'm glad that the snippet solves your problem. A possible bug I observe in the snippet is that `vim.wo.cc` is a string, so you have to convert it to a number before comparing it with the length of the longest line visible.

> I think it would be nice if this was possible to achieve with the options

As you can see in the implementation of this plugin, this plugin is implemented in a very tricky way and the logic has now become unnecessarily complicated due to a neovim bug neovim/neovim#23227 so I have no plan to further make it even more complex before the bug getting fixed (then I can cleanup the implementation). So if the snippet can solve your problem there is no plan to add an extra option. Please understand.

**Comment by Hubro** [comment]
Completely understandable :+1:

**Comment by Bekaboo** [comment]
Hi, please test commit 8f19d5f on branch `dev` if you are interested. This allows users to pass a callback function to `opts.scope`. I believe this is a more elegant way to achieve what you want.

**Comment by Hubro** [comment]
@Bekaboo Nice, that makes sense! My ideal setup would be something like:

```lua
scope = function()
  local visible_lines = vim.api.nvim_buf_get_lines(
    0, vim.fn.line("w0") - 1, vim.fn.line("w$"), false
  )
  local longest_visible_line =
    math.max(unpack(vim.tbl_map(vim.fn.strdisplaywidth, visible_lines)))

  if longest_visible_line > COLORCOLUMN then   -- ?
    return longest_visible_line
  else
    return vim.fn.strdisplaywidth(vim.api.nvim_get_current_line())
  end
end,
```

But I'm stuck on the COLORCOLUMN value, since this could vary per window.

I see there's a fairly chunky function `resolve_cc` in `autocmds.lua` that parses the colorcolumn value and returns a number, but it's not exported.

**Comment by Bekaboo** [comment]
Hi, please test commit dcb8165f, you should be able to access that function through `require('deadcolumn').utils.resolve_cc` or `require('deadcolumn.utils').resolve_cc`.

**Comment by Hubro** [comment]
It almost works perfectly! :grin: 

```lua
scope = function()
  local lines = vim.api.nvim_buf_get_lines(
    0,
    vim.fn.line("w0") - 1,
    vim.fn.line("w$"),
    false
  )
  local longest_on_screen_line =
    math.max(unpack(vim.tbl_map(vim.fn.strdisplaywidth, lines)))

  local cc = require("deadcolumn.utils").resolve_cc(vim.wo.cc)

  if cc ~= nil and longest_on_screen_line >= cc then
    return longest_on_screen_line
  else
    return vim.fn.strdisplaywidth(vim.api.nvim_get_current_line())
  end
end,
```

But there's a bug, for some reason the colorcolumn disappears shortly after I typing anything in insert mode:

https://user-images.githubusercontent.com/597206/235408938-b944580d-f4e5-4a1f-97b0-6c4e226cb750.mp4

**Comment by Bekaboo** [comment]
@Hubro Cannot reproduce. The following config works for me:

```lua
    opts = {
      modes = { 'i', 'R', 'c', 'n', 'v' },
      scope = function()
        local lines = vim.api.nvim_buf_get_lines(
          0,
          vim.fn.line('w0') - 1,
          vim.fn.line('w$'),
          false
        )
        local longest_on_screen_line =
          math.max(unpack(vim.tbl_map(vim.fn.strdisplaywidth, lines)))

        local cc = require('deadcolumn.utils').resolve_cc(vim.wo.cc)

        if cc ~= nil and longest_on_screen_line >= cc then
          return longest_on_screen_line
        else
          return vim.fn.strdisplaywidth(vim.api.nvim_get_current_line())
        end
      end,
    }
```

please provide full config if this still bugs you.

**Comment by Hubro** [comment]
@Bekaboo My `modes` was set to: `{ "i", "ic", "ix", "R", "Rc", "Rx", "Rv", "Rvc", "Rvx", "n" }`, which is the default value from the README, plus `"n"` because I want to see the columns in normal mode as well.

I changed the value to `{ 'i', 'R', 'c', 'n', 'v' }` as you suggested, and now it works as expected :confused: 

I have no clue what values like `"ic"`, `"ix"`, `"Rvc"` etc mean. Are these actual modes in Neovim or is it a Deadcolumn.nvim concept?

**Comment by Bekaboo** [comment]
@Bekaboo Cannot reproduce. This configuration works for me:

```lua
opts = {
  modes = { 'i', 'ic', 'ix', 'R', 'Rc', 'Rx', 'Rv', 'Rvc', 'Rvx', 'n' },
  scope = function()
    local lines = vim.api.nvim_buf_get_lines(
      0,
      vim.fn.line('w0') - 1,
      vim.fn.line('w$'),
      false
    )
    local longest_on_screen_line =
      math.max(unpack(vim.tbl_map(vim.fn.strdisplaywidth, lines)))

    local cc = require('deadcolumn.utils').resolve_cc(vim.wo.cc)

    if cc ~= nil and longest_on_screen_line >= cc then
      return longest_on_screen_line
    else
      return vim.fn.strdisplaywidth(vim.api.nvim_get_current_line())
    end
  end,
},
```

Make sure you are testing against the newest version on branch `dev` (commit 42e44ff2ec85f878942faf68b2a3ff631841697f)

**Comment by Bekaboo** [comment]
I closed this issue because it is a feature request, for bug report please open another issue and follow the format of the template, provide a minimal config and steps to reproduce.
